osu.Game.Rulesets.Osu/Mods/OsuModRelax.cs

```diff
@@ -31,32 +34,41 @@ public class OsuModRelax : ModRelax, IUpdatableByPlayfield, IApplicableToDrawabl
 
         private OsuInputManager osuInputManager;
 
+        private Score score;
+        private DrawableRuleset<OsuHitObject> drawableRuleset;
         private ReplayState<OsuAction> state;
         private double lastStateChangeTime;
+        private OsuFramedReplayInputHandler replayInputHandler;
+        private readonly BindableBool replayLoaded = new BindableBool();
 
-        private bool hasReplay;
+        public void ApplyToPlayer(Player player)
+        {
+            replayInputHandler = (OsuFramedReplayInputHandler)osuInputManager.ReplayInputHandler;
+            updateState();
+            osuInputManager.AllowUserPresses = false;
+        }
 
         public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
         {
             // grab the input manager for future use.
             osuInputManager = (OsuInputManager)drawableRuleset.KeyBindingInputManager;
+            replayLoaded.BindTo(drawableRuleset.HasReplayLoaded);
+            replayLoaded.BindValueChanged(_ => updateState());
+            this.drawableRuleset = drawableRuleset;
         }
 
-        public void ApplyToPlayer(Player player)
+        private void updateState()
         {
-            if (osuInputManager.ReplayInputHandler != null)
+            if (replayInputHandler != null)
             {
-                hasReplay = true;
-                return;
+                score = drawableRuleset.ReplayScore;
+                replayInputHandler.HandleActionInput = !(replayLoaded.Value && score.ScoreInfo.IsLegacyScore);
```

Comment:
This will cause lazer-made classic scores to use input from this mod instead of the replay frames.

How about instead clearing the actions of all replay frames at decode time if made with stable? There's already this condition to determine if an incoming score was made on stable:
https://github.com/ppy/osu/blob/52b9616909d57f514dcca68c8e21ac4e1de138c0/osu.Game/Scoring/Legacy/LegacyScoreDecoder.cs#L69

Then, this mod could check if `score.Replay.Frames.Any(f => (f as OsuReplayFrame)?.Actions.Any() == true);`

Reply:
What's the 0 action check about? stable still records key presses with relax.
![screenshot003](https://user-images.githubusercontent.com/50285552/125441533-b2981817-c688-46e2-8793-72ca9bc4d367.jpg)

Reply:
Still not sure on this one.

Maybe cc @peppy - tl;dr: there's no way to currently know if a replay is a stable-made replay from within the relax mod, and this PR is using `ModClassic` being present as the heuristic instead.

From discussions above, my suggestion was to detect stable replays at decode time and clear all actions, then switch on that in this code instead (indicating all frames with 0 actions == stable-made replay == relax mod does the hits even in replays).  
However that has the potential that, suppose, if a lazer-made classic-mod replay exists with 0 actions and the relax mod changes its leniency or whatever, it'll also perform the hits in the lazer-made replay.

I'm not sure there's a better way other than one of the two methods proposed - the existing code in this PR, or my own proposal.

Reply:
Is another potential solution to wait until we expand `ScoreInfo` to contain some kind of version specification, where we can determine whether a score was set on stable or lazer (regardless of `ModClassic`)? I don't see this issue as a super-high priority one, so it might be better to wait for such a solution if it sounds amicable.

Reply:
Yeah that was another thought going in my head, but I wasn't seeing the bigger picture. Shall we close this for the time being then?

Reply:
I'd say so. Not against temporary fixes but this one is really hard to warrant..